### DISCUSSION TITLE
请问如何实现主动发送消息？

### DISCUSSION THREAD

**Discussion by TianLuoRain** [mamoe/mirai#1187] [Q&A] [answered]
我看到
`Bot.getInstance(botQQ).getFriend(friendQQ).sendMessage(message);
但该方法好像不能用于插件体内？
在文档和讨论中也没有找到明确的相关内容。
我猜我大概又搞错了什么，但是我无法自己解决这个问题😫😫。

**Comment by Him188** [comment, marked as answer]
可以用

**Reply by TianLuoRain** [reply]
谢谢，我猜想的原因是，系统启动时先加载插件，再进行机器人登录。所以如果在onEnable()中直接写让机器人主动发送一条消息的语句
`Bot.getInstance(botQQ).getFriend(friendQQ).sendMessage(message);`
就会报错
`java.util.NoSuchElementException`
因为在此时还没有任何机器人登录。
我之后会尝试开启新线程并等待一段时间看看能否实现。

**Reply by Him188** [reply]
可以监听 `BotOnlineEvent` 即 Bot 上线

**Reply by TianLuoRain** [reply]
谢谢，监听`BotOnlineEvent`后成功了。
（虽然还是不知道该怎么做定时发送消息🤣🤣）

**Reply by Him188** [reply]
可以去了解如何用 Java 做定时任务, 每一段时间就用 `Bot.getInstances()` 获取在线 `Bot` 列表后对每个 `Bot` 依次发消息
